smbclient.c

```diff
@@ -1649,20 +1651,28 @@ PHP_FUNCTION(smbclient_getxattr)
 	if ((smbc_getxattr = smbc_getFunctionGetxattr(state->ctx)) == NULL) {
 		RETURN_FALSE;
 	}
-	/* TODO: 1000 chars should be enough for everyone...?
-	 * However, doing an initial blank call to determine the response size
-	 * seems wasteful, and vulnerable to a time-of-check, time-of-use
-	 * error. */
-	if ((retsize = smbc_getxattr(state->ctx, url, name, values, sizeof(values))) >= 0) {
-		if (retsize > sizeof(values)) {
-			retsize = sizeof(values);
-		}
-#if PHP_MAJOR_VERSION >= 7
-		RETURN_STRINGL(values, retsize);
-#else
-		RETURN_STRINGL(values, retsize, 1);
-#endif
+
+	xattr_size = smbc_getxattr(state->ctx, url, name, NULL, 0);
+
+	if(xattr_size < 0) {
+		goto fail;
 	}
+
+	if(xattr_size == 0) {
+		RETURN_EMPTY_STRING();
+	}
+
+	values = do_alloca(xattr_size + 1, use_heap);
```

Comment:
why not a simple emalloc ?

BTW, for PHP 7+, will be better to use a zend_string, which will avoid double allocation (RETURN_STR will use the allocated string, while RETURN_STRING allocate a new one)